base/applications/games/solitaire/solundo.cpp

```diff
@@ -0,0 +1,149 @@
+#include "solitaire.h"
+
+static int source_id = -1;
+static int destination_id = -1;
+static int number_of_cards = 0;
+static int prev_score = 0;
+static int prev_visible_pile_cards = 0;
+
+void SetUndo(
+    int set_source_id,
+    int set_destination_id,
+    int set_number_of_cards,
+    int set_prev_score,
+    int set_prev_visible_pile_cards)
+{
+    if (set_source_id == set_destination_id)
+        return;
+
+    source_id = set_source_id;
+    destination_id = set_destination_id;
+    number_of_cards = set_number_of_cards;
+    prev_score = set_prev_score;
+    prev_visible_pile_cards = set_prev_visible_pile_cards;
+    SetUndoMenuState(true);
+}
+
+void ClearUndo(void)
+{
+    source_id = -1;
+    destination_id = -1;
+    number_of_cards = 0;
+    SetUndoMenuState(false);
+}
+
+void Undo(void)
+{
+    CardRegion *source = NULL;
+    CardRegion *destination = NULL;
+
+    if (source_id < 1 || destination_id < 1 || number_of_cards < 1)
+    {
+        ClearUndo();
+        return;
+    }
+
+    if (source_id >= ROW_ID)
```

Comment:
Shouldn't source_id be tested for too large values? (if it turns out that e.g. `source_id - ROW_ID` is larger than the total number of items in `pRowStack`, or same comment for pSuitStack ...)

Reply:
Yes, of course. I added the check for large values above that code, thanks!

Reply:
Does the source_id also take into account the pSuitStack size?

Reply:
It does now, thanks!